awx/main/analytics/subsystem_metrics.py

```diff
@@ -0,0 +1,306 @@
+            FloatM('callback_receiver_events_insert_db_seconds',
+                   'Time spent saving events to database'),
+            IntM('callback_receiver_events_insert_db',
+                 'Number of events batch inserted into database'),
+            HistogramM('callback_receiver_batch_events_insert_db',
+                       'Number of events batch inserted into database',
+                       settings.SUBSYSTEM_METRICS_BATCH_INSERT_BUCKETS),
+            FloatM('subsystem_metrics_pipe_execute_seconds',
+                   'Time spent saving metrics to redis and sending data to other nodes'),
+            IntM('subsystem_metrics_pipe_execute_calls',
+                 'Number of calls to pipe_execute'),
+        ]
+        # turn metric list into dictionary with the metric name as a key
+        self.METRICS = {}
+        for m in METRICSLIST:
+            self.METRICS[m.field] = m
+
+        # track last time metrics were sent to other nodes
+        self.previous_send_metrics = SetFloatM('send_metrics_time',
+                                               'Timestamp of previous send_metrics call')
+
+    def clear_values(self):
+        for m in self.METRICS.values():
+            m.clear_value(self.conn)
+        self.metrics_have_changed = True
+        self.conn.delete(root_key + "_lock")
+
+    def inc(self, field, value):
+        if value != 0:
+            self.METRICS[field].inc(value)
+            self.metrics_have_changed = True
+            if self.auto_pipe_execute is True and self.should_pipe_execute() is True:
+                self.pipe_execute()
+
+    def set(self, field, value):
+        self.METRICS[field].set(value)
+        self.metrics_have_changed = True
+        if self.auto_pipe_execute is True and self.should_pipe_execute() is True:
+            self.pipe_execute()
+
+    def observe(self, field, value):
+        self.METRICS[field].observe(value)
+        self.metrics_have_changed = True
+        if self.auto_pipe_execute is True and self.should_pipe_execute() is True:
+            self.pipe_execute()
+
+    def serialize_local_metrics(self):
+        data = self.load_local_metrics()
+        return json.dumps(data)
+
+    def load_local_metrics(self):
+        # generate python dictionary of key values from metrics stored in redis
+        data = {}
+        for field in self.METRICS:
+            data[field] = self.METRICS[field].decode(self.conn)
+        return data
+
+    def store_metrics(self, data_json):
+        # called when receiving metrics from other instances
+        data = json.loads(data_json)
+        if self.instance_name != data['instance']:
+            logger.debug(f"{self.instance_name} received subsystem metrics from {data['instance']}")
+        self.conn.set(root_key + "_instance_" + data['instance'], data['metrics'])
+
+    def should_pipe_execute(self):
+        if time.time() - self.last_pipe_execute > self.pipe_execute_interval:
```

Comment:
I don't love using a timeout as a solution to reduce the number of metrics request to redis. We can have no loss in performance if we piggyback with pipelining onto the redis callback receiver pop()

Reply:
Maybe this is OK. Since these are atomic incr and dec operations.

I'm not sure if this pattern for pipelining will work well when used in a process that isn't a long-lived one that continuously loops i.e. the task manager. In that case, we want to use pipelining to form a single transaction and "commit" it at the end of each task manager run; but we can cross that bridge when we get to it.

Reply:
@chrismeyersfsu
> We can have no loss in performance if we piggyback with pipelining onto the redis callback receiver pop()

I found performance loss when introducing pipelining on every `read()` loop. I outlined the details here, but I think it got lost in the "hidden" section of this PR
https://github.com/ansible/awx/pull/9461#issuecomment-800580082

you can run this code snippet in awx-manage shell to get an idea of the overhead

```python
import time
import redis
e = "======================================= this is an event ================================================================"
connection = redis.Redis.from_url(settings.BROKER_URL)
connection_pipe = redis.Redis.from_url(settings.BROKER_URL).pipeline()
connection.flushdb()
for _ in range(100000):
    connection.rpush("sbf_test", e)
x = time.time()
for _ in range(100000):
    connection_pipe.blpop("sbf_test")
    connection_pipe.llen("sbf_test")
    for _ in range(10):
        connection_pipe.incr("sbf_inc", 1)
    connection_pipe.execute()
print("pipe pop, len, inc: ", time.time() - x)
connection.flushdb()
for _ in range(100000):
    connection.rpush("sbf_test", e)
x = time.time()
for _ in range(100000):
    connection.blpop("sbf_test")
print("no pipe just pop: ", time.time() - x)
connection.flushdb()
for _ in range(100000):
    connection.rpush("sbf_test", e)
x = time.time()
for _ in range(100000):
    connection_pipe.blpop("sbf_test")
    connection_pipe.execute()
print("pipe just pop: ", time.time() - x)```

```
pipe pop, len, inc:  19.783135890960693
no pipe just pop:  4.3485107421875
pipe just pop:  6.442138433456421
```

I do agree that for task manager we'll want a way to do a single transaction at the end. Currently in this PR, `inc` and `set` only modify object memory values -- they are not doing pipeline operations. So in the task manager we could just accumulate the values in a Metrics() object, then do a `pipe_execute()` whenever we wish at the end.